[Jonas Goh] iP

jonasgwt opened the pull request:
# DukePro
> “Your mind is for having ideas, not holding them.” – David Allen ([source](https://dansilvestre.com/productivity-quotes))

DukePro frees your mind of having to remember things you need to do. It's,
- text-based
- easy to learn
- FAST SUPER FAST to use

All you need to do is,

1. download it from [here](https://nus-cs2103-ay2223s1.github.io/website/schedule/week4/project.html).
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you 😉

And it is **FREE!**

Features:

- [x] Managing tasks
- [ ] Managing deadlines (coming soon)
- [ ] Reminders (coming soon)

If you Java programmer, you can use it to practice Java too. Here's the `main` method:

```
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

jonasgwt committed:
Completed Level 1

jonasgwt committed:
Completed Level 2

jonasgwt committed:
Completed Level 3

jonasgwt committed:
Added new classes and reorganised Duke.reply

jonasgwt committed:
Completed Level 4

jonasgwt committed:
Added Testing

jonasgwt committed:
Fixed Bug in task marking

jonasgwt committed:
Added More Tests

jonasgwt committed:
Added DukeException

jonasgwt committed:
Added Exception Handling

jonasgwt committed:
Updated Tests

jonasgwt committed:
Added Delete Feature

jonasgwt committed:
Updated Tests

jonasgwt committed:
Added Enums

jonasgwt committed:
Added child classes for DukeException and updated test

jonasgwt committed:
Add Save Feature

jonasgwt committed:
Change to save date information in LocalDate type

jonasgwt committed:
Merge branch 'branch-Level-7'

jonasgwt committed:
Merge branch 'branch-Level-8'

jonasgwt committed:
Tasklist class: fix bug when parsing date
Deadline class: fix missing imports
DukeException package: organise DukeExcpetions into package

jonasgwt committed:
duke.exception package: change naming

jonasgwt committed:
Add more OOP

Majority of the program is handled by Duke class

Closely related code can be extracted out as classes.

Let's,
* Extract Ui class that deals with interactions with the user
* Extract Storage class that deals with loading tasks from the file
and saving tasks in the file
* Extract Parser class that deals with making sense of user command
* Extract TaskList class that contains the task list
* Extract *Command classes that handles the execution of user command

jonasgwt committed:
Organize into Packages

jonasgwt committed:
Merge branch 'add-gradle-support'

jonasgwt committed:
Add JUnit Tests

jonasgwt committed:
UnknownCommand class: Error message fix

jonasgwt committed:
Add JavaDoc comments

jonasgwt committed:
Tweak the code to comply with a coding standard

jonasgwt committed:
Add find feature and updated tests

jonasgwt committed:
Merge branch 'branch-A-JavaDoc'

jonasgwt committed:
Merge branch 'branch-A-CodingStandard'

jonasgwt committed:
Merge branch 'branch-Level-9'

jonasgwt committed:
Fix Merge Conflicts

jonasgwt committed:
Fix code style issues

jonasgwt committed:
Add Ui for Duke

jonasgwt committed:
Merge branch 'branch-Level-10'

Darren12345677 commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,40 @@
+ * The class that encapsulates the Duke bot.
+ */
+public class Duke {
+    /** Stores all the data of tasks */
+    private final TaskList data;
+    /** Deals with loading tasks and saving tasks */
+    private Storage storage;
+    /** Deals with making sense of the user command */
+    private final Parser parser;
```
I like how you even explain your private attributes👍

Darren12345677 commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,98 @@
+package duke;
+
+import java.time.LocalDate;
+import java.time.format.DateTimeParseException;
+
+import duke.command.AddCommand;
+import duke.command.Command;
+import duke.command.DeleteCommand;
+import duke.command.ExitCommand;
+import duke.command.FindCommand;
+import duke.command.ListCommand;
+import duke.command.MarkCommand;
+import duke.exception.DukeException;
+import duke.exception.InvalidInput;
+import duke.exception.UnknownCommand;
```
I like the way you categorise your imports, it looks really neat

Darren12345677 commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,98 @@
+        } catch (IllegalArgumentException e) {
+            throw new UnknownCommand();
+        }
+        switch (command) {
```
Great use of switch statements 😃

Darren12345677 commented on src/main/java/duke/command/AddCommand.java:
```diff
@@ -0,0 +1,88 @@
+        }
+        tasks.add(task);
+        storage.saveData(tasks);
+        return "Got it. I've added this task:\n" + task + "\nNow you have " + tasks.size() + " tasks.";
```
Since there are many operators, could line wrapping be used here to improve readability?

Darren12345677 commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,106 @@
+ * to file.
+ */
+public class Storage {
+    /**
+     * Class constructor that initialises the path where we will
+     * save the data in.
+     *
+     * @throws DukeException If we are unable to create the 'data'
+     *                       directory.
+     */
```
I think it is really important to explain why exceptions were thrown. Nice job here!

Darren12345677 reviewed (commented):
Looks good.👍 
Overall, I found your code easy to read for the most part. JavaDocs were really detailed and appropriate.

izzahaj commented on src/main/java/duke/exception/UnknownCommand.java:
```diff
@@ -0,0 +1,14 @@
+    public UnknownCommand() {
+        super("Unknown command!");
+    }
+}
```
Great use of child exception classes to handle more specific exceptions!

izzahaj commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,138 @@
+     * @throws InvalidInput If the string given is not a number or the number is
+     *                      less than 0 or more than the size of the ArrayList.
+     */
+    public int stringIndexToInt(String res) throws InvalidInput {
```
Perhaps prepending this method with an appropriate verb (e.g. 'get' or 'convert') would be better?

izzahaj commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,138 @@
+    }
+
+    /**
+     * Returns the tasks saved in the provided index.
```
Perhaps using `task` instead of `tasks` would be more appropriate? Since this method only returns a single Task object.

izzahaj commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,138 @@
+     * @param index The index of the tasks of interest.
+     * @return The task of interest.
+     */
+    public Task get(int index) {
```
Perhaps renaming the method to `getTask` would be clearer?

izzahaj reviewed (commented):
Overall, I found your code organised and readable. You also adhered to the coding standards very well, apart from a few minor improvements you could make.

jonasgwt committed:
Add Tweaks to UI

jonasgwt committed:
Fix bug with scaling of DialogBox

jonasgwt committed:
Add use of assertions

There was no use of assertions.

Assertions can define assumptions about the program state so that
runtime can verify them, reducing the risk of bugs.

Let's,
* Add assertions to verify preconditions
* Add assertions to verify postconditions
* Add assertions to verify control-flow invariants

jonasgwt committed:
Refactor code to improve code quality

Code did not comply with some coding best practices.

Ensured code complied with coding best practices.

Let's,
* Refactor code to comply with the SLAP
* Refactor code to avoid the use of magic numbers
* Refactor code to avoid complicated expressions
* Refactor code to avoid deep nesting
* Added java docs for some methods
* Avoid the use of empty catch blocks
* Minimise code duplication

jonasgwt committed:
Merge pull request #1 from jonasgwt/branch-A-Assertions

Add use of assertions

jonasgwt committed:
Merge branch 'master' into branch-A-CodeQuality

jonasgwt committed:
Merge pull request #2 from jonasgwt/branch-A-CodeQuality

Refactor code to improve code quality

jonasgwt committed:
Set up CI

jonasgwt committed:
Add Checkstyle

jonasgwt committed:
Add sorting for tasks

jonasgwt committed:
Merge branch 'branch-C-Sort'

jonasgwt committed:
Remove assertions for argument checking in public methods

jonasgwt committed:
Remove assertions for argument checking in public methods

jonasgwt committed:
Add Help command

jonasgwt committed:
Merge branch 'branch-C-Help'

jonasgwt committed:
Alternate text alignment between duke and user

jonasgwt committed:
Change ExitCommand such that it exits the application on "bye"

jonasgwt committed:
Add User Guide

jonasgwt committed:
Add screenshot

jonasgwt committed:
Fix JavaDoc comments
